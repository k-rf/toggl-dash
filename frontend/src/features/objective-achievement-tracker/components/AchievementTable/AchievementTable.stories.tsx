import type { Meta, StoryObj } from "@storybook/react";

import { AchievementTable, MonthlyObjectiveInfo } from "./AchievementTable";

export default { component: AchievementTable } as Meta<typeof AchievementTable>;

type Story = StoryObj<typeof AchievementTable>;

const data: MonthlyObjectiveInfo[] = [
  {
    month: 1,
    achievementTime: 100,
    availableTime: 120,
    borderTime: 80,
  },
  {
    month: 2,
    achievementTime: 90,
    availableTime: 110,
    borderTime: 90,
  },
  {
    month: 3,
    achievementTime: 100,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 4,
    achievementTime: 60,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 5,
    achievementTime: 100,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 6,
    achievementTime: 100,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 7,
    achievementTime: 100,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 8,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 9,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 10,
    availableTime: 120,
    borderTime: 70,
  },
  {
    month: 11,
    availableTime: 120,
    borderTime: 130,
  },
  {
    month: 12,
    availableTime: 120,
    borderTime: 70,
  },
];

export const Default: Story = {
  args: { data },
};
