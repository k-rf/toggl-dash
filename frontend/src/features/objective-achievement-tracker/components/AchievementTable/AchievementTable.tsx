import { Box, SxProps, Theme } from "@mui/material";
import {
  createColumnHelper,
  flexRender,
  getCoreRowModel,
  noop,
  useReactTable,
} from "@tanstack/react-table";
import { FC } from "react";
import { match } from "ts-pattern";

import {
  ElTable,
  ElTableBody,
  ElTableCell,
  ElTableHead,
  ElTableRow,
} from "~/components/Elements/ElTable";

export type MonthlyObjectiveInfo = {
  month: number;
  borderTime: number;
  availableTime: number;
  achievementTime?: number;
};

const columnHelper = createColumnHelper<MonthlyObjectiveInfo>();

const defaultColumns = [
  columnHelper.accessor("month", {
    header: noop,
    cell: (info) => {
      return <strong>{info.getValue()} 月</strong>;
    },
  }),
  columnHelper.accessor("availableTime", {
    header: () => <span>利用可能</span>,
    cell: (info) => <span>{info.getValue()} H</span>,
  }),
  columnHelper.accessor("borderTime", {
    header: () => <span>ボーダー</span>,
    cell: (info) => <span>{info.getValue()} H</span>,
  }),
  columnHelper.accessor("achievementTime", {
    header: () => <span>実績</span>,
    cell: (info) => {
      const borderTime = info.row.getValue<number>("borderTime");
      const achievementTime = info.getValue();

      const style = match<boolean, SxProps<Theme>>(borderTime <= (achievementTime ?? 0))
        .with(true, () => {
          return {
            color: (theme) => theme.palette.success.light,
            fontWeight: 700,
          };
        })
        .with(false, () => {
          return {
            color: (theme) => theme.palette.error.main,
          };
        })
        .exhaustive();

      return (
        <Box component="span" sx={style}>
          {achievementTime && `${achievementTime} H`}
        </Box>
      );
    },
  }),
];

type Props = {
  data: MonthlyObjectiveInfo[];
};

export const AchievementTable: FC<Props> = (props) => {
  const table = useReactTable({
    data: props.data,
    columns: defaultColumns,
    getCoreRowModel: getCoreRowModel(),
  });

  return (
    <Box width={1000}>
      <ElTable size="small" sx={{ writingMode: "vertical-lr" }}>
        <ElTableHead>
          {table.getHeaderGroups().map((headerGroup) => (
            <ElTableRow key={headerGroup.id}>
              {headerGroup.headers.map((header) => (
                <ElTableCell
                  key={header.id}
                  colSpan={header.colSpan}
                  sx={{ writingMode: "horizontal-tb" }}
                >
                  {header.isPlaceholder
                    ? null
                    : flexRender(header.column.columnDef.header, header.getContext())}
                </ElTableCell>
              ))}
            </ElTableRow>
          ))}
        </ElTableHead>
        <ElTableBody>
          {table.getRowModel().rows.map((row) => (
            <ElTableRow key={row.id}>
              {row.getVisibleCells().map((cell) => (
                <ElTableCell key={cell.id} sx={{ writingMode: "horizontal-tb" }}>
                  {flexRender(cell.column.columnDef.cell, cell.getContext())}
                </ElTableCell>
              ))}
            </ElTableRow>
          ))}
        </ElTableBody>
      </ElTable>
    </Box>
  );
};
