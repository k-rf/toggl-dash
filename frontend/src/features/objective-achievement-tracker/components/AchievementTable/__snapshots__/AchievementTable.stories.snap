// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/Objective-achievement-tracker/Components/AchievementTable/AchievementTable > Default 1`] = `
<body
  style=""
>
  <div>
    <div
      class="MuiBox-root css-9hu0cu"
    >
      <table
        class="MuiTable-root css-1avgbhz-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-18gyjk9-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-j2zpdo-MuiTableCell-root"
              colspan="1"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-j2zpdo-MuiTableCell-root"
              colspan="1"
              scope="col"
            >
              <span>
                利用可能
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-j2zpdo-MuiTableCell-root"
              colspan="1"
              scope="col"
            >
              <span>
                ボーダー
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-j2zpdo-MuiTableCell-root"
              colspan="1"
              scope="col"
            >
              <span>
                実績
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                1
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                80
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                100 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                2
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                110
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                90
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                90 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                3
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                100 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                4
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              >
                60 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                5
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                100 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                6
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                100 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                7
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1ayqwaf"
              >
                100 H
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                8
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              />
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                9
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              />
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                10
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              />
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                11
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                130
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              />
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-18gyjk9-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <strong>
                12
                 月
              </strong>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                120
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span>
                70
                 H
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1lqgvws-MuiTableCell-root"
            >
              <span
                class="MuiBox-root css-1bv7w9j"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
